spmd/api.py

```diff
@@ -0,0 +1,329 @@
+                                # been prepared in value_remap
+                                value_remap[dtn] = a
+                            elif dtn.op == "output":
+                                assert (
+                                    len(dtn.args) == 1 and len(dtn.args[0]) == 1
+                                ), f"Expecting single output, but got {dtn.args}"
+                                #a.replace_all_uses_with(value_remap[dtn.args[0][0]])
+                                new_args.append(value_remap[dtn.args[0][0]])
+                            else:
+                                if dtn.op == "get_attr":
+                                    setattr(gm, dtn.target, getattr(traced_dispatch, dtn.target))
+                                with gm.graph.inserting_before(node):
+                                    value_remap[dtn] = gm.graph.node_copy(
+                                        dtn, lambda n: value_remap[n]
+                                    )
+
+                gm.graph.erase_node(node)
+                gm.graph.output(new_args)
+                break
+            else:
+                raise ValueError(f"Unrecognized node {node}")
+
+        #gm.graph.print_tabular()
+
+        # replace nodes in local traced graph with DTensor's dispatch graph
+        for node in gm.graph.nodes:
+            if node not in replacements:
+                continue
+
+            traced_dispatch = replacements[node]
+            # Map DT's dispatch graph input placeholder nodes to the ones in
+            # local traced graph. It uses index-based accessing, which is
+            # brittle, just for testing purpose.
+            flatten_args, _ = tree_flatten(node.args)
+            i, value_remap = 0, {}
+            for dtn in traced_dispatch.graph.nodes:
+                if dtn.op == "placeholder":
+                    value_remap[dtn] = flatten_args[i]
+                    i += 1
+
+            # insert DT's dispatch graph to traced local graph.
+            with gm.graph.inserting_before(node):
+                for dtn in traced_dispatch.graph.nodes:
+                    if dtn.op == "placeholder":
+                        # do nothing, ignore placeholders, as it has already
+                        # been prepared in value_remap
+                        pass
+                    elif dtn.op == "output":
+                        assert (
+                            len(dtn.args) == 1 and len(dtn.args[0]) == 1
+                        ), f"Expecting single output, but got {dtn.args}"
+                        node.replace_all_uses_with(value_remap[dtn.args[0][0]])
+                    else:
+                        value_remap[dtn] = gm.graph.node_copy(
+                            dtn, lambda n: value_remap[n]
+                        )
+
+
+        gm.graph.lint()
+        gm.graph.eliminate_dead_code()
+        if dist.get_rank() == 0:
+            gm.graph.print_tabular()
```

Comment:
Hey @Chillee, the returned backward graph looks correct to me, but when I call backward, it does not launch allreduce in the graph. Do you know why?

```
opcode         name                  target                                          args                                                         kwargs          
-------------  --------------------  ----------------------------------------------  -----------------------------------------------------------  --------        
placeholder    t_1                   t_1                                             ()                                                           {}              
placeholder    primals_5             primals_5                                       ()                                                           {}              
placeholder    addmm                 addmm                                           ()                                                           {}              
placeholder    tangents_1            tangents_1                                      ()                                                           {}              
call_function  t                     aten.t.default                                  (t_1,)                                                       {}              
call_function  mm_3                  aten.mm.default                                 (tangents_1, t)                                              {}              
call_function  t_9                   aten.t.default                                  (tangents_1,)                                                {}              
call_function  mm_4                  aten.mm.default                                 (t_9, addmm)    
call_function  t_10                  aten.t.default                                  (mm_4,)                                                      {}     [73/1873]
call_function  sum_3                 aten.sum.dim_IntList                            (tangents_1, [0], True)                                      {}              
call_function  view_2                aten.view.default                               (sum_3, [10])                                                {}              
call_function  t_11                  aten.t.default                                  (t_10,)                                                      {}              
call_function  t_12                  aten.t.default                                  (mm_3,)                                                      {}              
call_function  mm_5                  aten.mm.default                                 (t_12, primals_5)                                            {}              
call_function  t_13                  aten.t.default                                  (mm_5,)                                                      {}              
call_function  sum_4                 aten.sum.dim_IntList                            (mm_3, [0], True)                                            {}              
call_function  view_3                aten.view.default                               (sum_4, [10])                                                {}              
call_function  t_14                  aten.t.default                                  (t_13,)                                                      {}              
call_function  clone                 aten.clone.default                              (t_14,)                                                      {}              
get_attr       _tensor_constant0     _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1     _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default    c10d.allreduce_.default                         ([clone], _tensor_constant0, _tensor_constant1, -1)          {}              
call_function  comm_result           <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default,)                                        {}              
call_function  getitem               <built-in function getitem>                     (comm_result, 0)                                             {}              
call_function  getitem_1             <built-in function getitem>                     (getitem, 0)                                                 {}              
call_function  wait_comm             <function _wait_comm at 0x7f1e55865b80>         (getitem_1,)                                                 {}              
call_function  clone_1               aten.clone.default                              (view_3,)                                                    {}              
get_attr       _tensor_constant0_1   _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1_1   _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default_1  c10d.allreduce_.default                         ([clone_1], _tensor_constant0_1, _tensor_constant1_1, -1)    {}              
call_function  comm_result_1         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_1,)                                      {}              
call_function  getitem_2             <built-in function getitem>                     (comm_result_1, 0)                                           {}              
call_function  getitem_3             <built-in function getitem>                     (getitem_2, 0)                                               {}              
call_function  wait_comm_1           <function _wait_comm at 0x7f1e55865b80>         (getitem_3,)                                                 {}              
call_function  clone_2               aten.clone.default                              (t_11,)                                                      {}              
get_attr       _tensor_constant0_2   _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1_2   _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default_2  c10d.allreduce_.default                         ([clone_2], _tensor_constant0_2, _tensor_constant1_2, -1)    {}              
call_function  comm_result_2         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_2,)                                      {}              
call_function  getitem_4             <built-in function getitem>                     (comm_result_2, 0)                                           {}              
call_function  getitem_5             <built-in function getitem>                     (getitem_4, 0)                                               {}              
call_function  wait_comm_2           <function _wait_comm at 0x7f1e55865b80>         (getitem_5,)                                                 {}
call_function  clone_3               aten.clone.default                              (view_2,)                                                    {}
get_attr       _tensor_constant0_3   _tensor_constant0                               ()                                                           {}
get_attr       _tensor_constant1_3   _tensor_constant1                               ()                                                           {}
call_function  allreduce__default_3  c10d.allreduce_.default                         ([clone_3], _tensor_constant0_3, _tensor_constant1_3, -1)    {}
call_function  comm_result_3         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_3,)                                      {}
call_function  getitem_6             <built-in function getitem>                     (comm_result_3, 0)                                           {}
call_function  getitem_7             <built-in function getitem>                     (getitem_6, 0)                                               {}
call_function  wait_comm_3           <function _wait_comm at 0x7f1e55865b80>         (getitem_7,)                                                 {}
output         output_1              output                                          ([wait_comm, wait_comm_1, wait_comm_2, wait_comm_3, None],)  {}
```

Reply:
I am using functorch in pytorch/pytorch nightly `1.13.0.dev20220915+cu113`